core/src/main/scala/caliban/interop/jsoniter/jsoniter.scala

```diff
@@ -78,20 +80,44 @@ private[caliban] object ValueJsoniter {
     out.writeObjectEnd()
   }
 
-  private def encodeResponseValue(x: ResponseValue, out: JsonWriter, depth: Int): Unit = x match {
-    case StringValue(value)                 => out.writeVal(value)
-    case BooleanValue(value)                => out.writeVal(value)
-    case IntValue.IntNumber(value)          => out.writeVal(value)
-    case IntValue.LongNumber(value)         => out.writeVal(value)
-    case IntValue.BigIntNumber(value)       => out.writeVal(value)
-    case FloatValue.FloatNumber(value)      => out.writeVal(value)
-    case FloatValue.DoubleNumber(value)     => out.writeVal(value)
-    case FloatValue.BigDecimalNumber(value) => out.writeVal(value)
-    case NullValue                          => out.writeNull()
-    case EnumValue(value)                   => out.writeVal(value)
-    case ResponseValue.ObjectValue(o)       => writeResponseObject(o, out, decrDepth(depth, out))
-    case ResponseValue.ListValue(l)         => writeResponseArray(l, out, decrDepth(depth, out))
-    case s: ResponseValue.StreamValue       => out.writeVal(s.toString)
+  private def encodeResponseValue(x: ResponseValue, out: JsonWriter, depth: Int): Unit = {
+
+    /**
+     * Note on performance:
+     *
+     * We keep the number of cases low (less than 12) for the most commonly used types and avoid using unapply methods in the pattern-matching.
+     * This way we can better utilize JVM inlining, while still preserving match safety via the compiler if we ever extend ResponseValue with more cases.
+     *
+     * If our value is not one of the common ones, we fallback to the full encoder.
+     */
+    def fullEncoder(): Unit = x match {
+      case v: StringValue                 => out.writeVal(v.value)
+      case v: BooleanValue                => out.writeVal(v.value)
+      case v: IntValue.IntNumber          => out.writeVal(v.value)
+      case v: IntValue.LongNumber         => out.writeVal(v.value)
+      case v: IntValue.BigIntNumber       => out.writeVal(v.value)
+      case v: FloatValue.FloatNumber      => out.writeVal(v.value)
+      case v: FloatValue.DoubleNumber     => out.writeVal(v.value)
```

Comment:
Shouldn’t we put the cases that are _not_ included in the fast encoder first? Or does it not make a difference to the compiler?

Reply:
I ended up only handling `FloatValue` in a separate method, seems that the number of cases is still low enough to get the ~20% improvement